### PR TITLE
Add account suspension features

### PR DESCRIPTION
**Pull request by robjloranger**
This renders all requests for that user's posts, collections and related
ActivityPub endpoints with 404 responses.

While suspended, users may not create or edit posts or collections. A 403
is returned to the authenticated user.

User status is listed in the admin user page

Admin view of user details shows status and has a button to activate or
suspend a user.

Resolves [T661](https://writefreely.org/tasks/661).

---

- [x] I have signed the [CLA](https://phabricator.write.as/L1)

### CODE REVIEW

File: migrations/v3.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * it under the terms of the GNU Affero General Public License, included
+ * in the LICENSE file in this source code package.
+ */
+
+package migrations
+
+func supportUserSuspension(db *datastore) error {
+	t, err := db.Begin()
+
+	_, err = t.Exec(`ALTER TABLE users ADD COLUMN suspended ` + db.typeBool() + ` DEFAULT '0' NOT NULL`)
```
Comment (thebaer) [inline migrations/v3.go]: Need to change this to be `status` integer column instead.

File: activitypub.go [hunk 2]
```diff
@@ -158,6 +174,14 @@ func handleFetchCollectionFollowers(app *App, w http.ResponseWriter, r *http.Req
 	if err != nil {
 		return err
 	}
+	suspended, err := app.db.IsUserSuspended(c.OwnerID)
+	if err != nil {
+		log.Error("fetch collection inbox: get owner: %v", err)
```
Comment (thebaer) [inline activitypub.go]: Should this error message say "get owner"?

File: activitypub.go [hunk 3]
```diff
@@ -204,6 +228,14 @@ func handleFetchCollectionFollowing(app *App, w http.ResponseWriter, r *http.Req
...
```
Comment (thebaer) [inline activitypub.go]: Should this error message say "get owner"?

File: activitypub.go [hunk 4]
```diff
@@ -105,6 +113,14 @@ func handleFetchCollectionOutbox(app *App, w http.ResponseWriter, r *http.Reques
...
```
Comment (thebaer) [inline activitypub.go]: Should this error message say "get owner"?

File: activitypub.go [hunk 5]
```diff
@@ -80,6 +80,14 @@ func handleFetchCollectionActivities(app *App, w http.ResponseWriter, r *http.Re
...
```
Comment (thebaer) [inline activitypub.go]: Should this error message say "get owner"?

File: activitypub.go [hunk 6]
```diff
@@ -238,6 +270,14 @@ func handleFetchCollectionInbox(app *App, w http.ResponseWriter, r *http.Request
 		// TODO: return Reject?
 		return err
 	}
+	suspended, err := app.db.IsUserSuspended(c.OwnerID)
+	if err != nil {
+		log.Error("fetch collection inbox: get owner: %v", err)
```
Comment (thebaer) [inline activitypub.go]: Should this error message say "get owner"?

File: collections.go [hunk 7]
```diff
@@ -395,6 +396,14 @@ func newCollection(app *App, w http.ResponseWriter, r *http.Request) error {
 		}
 		userID = u.ID
 	}
+	suspended, err := app.db.IsUserSuspended(userID)
+	if err != nil {
+		log.Error("new collection: get user: %v", err)
```
Comment (thebaer) [inline collections.go]: Should this error message say "get user"?

File: collections.go [hunk 8]
```diff
@@ -724,6 +733,15 @@ func handleViewCollection(app *App, w http.ResponseWriter, r *http.Request) erro
 		return err
 	}
 
+	suspended, err := app.db.IsUserSuspended(c.OwnerID)
+	if err != nil {
+		log.Error("view collection: get owner: %v", err)
```
Comment (thebaer) [inline collections.go]: Should this error message say "get owner"?

File: collections.go [hunk 9]
```diff
@@ -916,7 +938,6 @@ func existingCollection(app *App, w http.ResponseWriter, r *http.Request) error
 	if reqJSON && !isWeb {
 		// Ensure an access token was given
 		accessToken := r.Header.Get("Authorization")
-		u = &User{}
```
Comment (thebaer) [inline collections.go]: Won't the next line throw a `nil` error without this line?

File: migrations/migrations.go [hunk 10]
```diff
@@ -57,6 +58,7 @@ func (m *migration) Migrate(db *datastore) error {
 var migrations = []Migration{
 	New("support user invites", supportUserInvites),             // -> V1 (v0.8.0)
 	New("support dynamic instance pages", supportInstancePages), // V1 -> V2 (v0.9.0)
+	New("support users suspension", supportUserSuspension),      // V2 -> V3 ()
```
Comment (thebaer) [inline migrations/migrations.go]: We'll definitely get this into v0.11, so feel free to add that to the comment.

File: posts.go [hunk 11]
```diff
@@ -380,6 +380,16 @@ func handleViewPost(app *App, w http.ResponseWriter, r *http.Request) error {
 		}
 	}
 
+	suspended, err := app.db.IsUserSuspended(ownerID.Int64)
+	if err != nil {
+		log.Error("view post: get collection owner: %v", err)
```
Comment (thebaer) [inline posts.go]: Should this error message say "get collection owner"?

File: posts.go [hunk 12]
```diff
@@ -496,6 +506,15 @@ func newPost(app *App, w http.ResponseWriter, r *http.Request) error {
 	} else {
 		userID = app.db.GetUserID(accessToken)
 	}
+	suspended, err := app.db.IsUserSuspended(userID)
+	if err != nil {
+		log.Error("new post: get user: %v", err)
```
Comment (thebaer) [inline posts.go]: Should this error message say "get user"?

File: posts.go [hunk 13]
```diff
@@ -662,6 +680,15 @@ func existingPost(app *App, w http.ResponseWriter, r *http.Request) error {
 		}
 	}
 
+	suspended, err := app.db.IsUserSuspended(userID)
+	if err != nil {
+		log.Error("existing post: get user: %v", err)
```
Comment (thebaer) [inline posts.go]: Should this error message say "get user"?

File: posts.go [hunk 14]
```diff
@@ -856,11 +883,20 @@ func addPost(app *App, w http.ResponseWriter, r *http.Request) error {
 		ownerID = u.ID
 	}
 
+	suspended, err := app.db.IsUserSuspended(ownerID)
+	if err != nil {
+		log.Error("add post: get user: %v", err)
```
Comment (thebaer) [inline posts.go]: Should this error message say "get user"?

File: posts.go [hunk 15]
```diff
@@ -950,13 +986,22 @@ func pinPost(app *App, w http.ResponseWriter, r *http.Request) error {
 		userID = u.ID
 	}
 
+	suspended, err := app.db.IsUserSuspended(userID)
+	if err != nil {
+		log.Error("pin post: get user: %v", err)
```
Comment (thebaer) [inline posts.go]: Should this error message say "get user"?

File: posts.go [hunk 16]
```diff
@@ -1007,12 +1053,22 @@ func fetchPost(app *App, w http.ResponseWriter, r *http.Request) error {
...
+		ownerID = coll.OwnerID
 	}
 
 	p, err := app.db.GetPost(vars["post"], collID)
 	if err != nil {
 		return err
 	}
+	suspended, err := app.db.IsUserSuspended(ownerID)
+	if err != nil {
+		log.Error("fetch post: get owner: %v", err)
```
Comment (thebaer) [inline posts.go]: Should this error message say "get owner"?

File: posts.go [hunk 17]
```diff
@@ -1270,6 +1326,15 @@ func viewCollectionPost(app *App, w http.ResponseWriter, r *http.Request) error
 	}
 	c.hostName = app.cfg.App.Host
 
+	suspended, err := app.db.IsUserSuspended(c.OwnerID)
+	if err != nil {
+		log.Error("view collection post: get owner: %v", err)
```
Comment (thebaer) [inline posts.go]: Should this error message say "get owner"?

File: routes.go [hunk 18]
```diff
@@ -143,6 +144,7 @@ func InitRoutes(apper Apper, r *mux.Router) *mux.Router {
 	write.HandleFunc("/admin", handler.Admin(handleViewAdminDash)).Methods("GET")
 	write.HandleFunc("/admin/users", handler.Admin(handleViewAdminUsers)).Methods("GET")
 	write.HandleFunc("/admin/user/{username}", handler.Admin(handleViewAdminUser)).Methods("GET")
+	write.HandleFunc("/admin/user/{username}", handler.Admin(handleAdminToggleUserSuspended)).Methods("POST")
```
Comment (thebaer) [inline routes.go]: What do you think about making this `/admin/user/{username}/suspend` instead? As we might want to use `/admin/user/{username}` for other functions in the future.

File: schema.sql [hunk 19]
```diff
@@ -225,6 +225,7 @@ CREATE TABLE IF NOT EXISTS `users` (
   `password` char(60) CHARACTER SET latin1 COLLATE latin1_bin NOT NULL,
   `email` varbinary(255) DEFAULT NULL,
   `created` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `suspended` tinyint(1) NOT NULL DEFAULT 0,
```
Comment (thebaer) [inline schema.sql]: No need to update schema files; migrations take care of this.

File: sqlite.sql [hunk 20]
```diff
@@ -214,7 +214,8 @@ CREATE TABLE IF NOT EXISTS `users` (
   username TEXT NOT NULL UNIQUE,
   password TEXT NOT NULL,
   email TEXT DEFAULT NULL,
-  created DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
+  created DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  suspended INTEGER NOT NULL DEFAULT 0
```
Comment (thebaer) [inline sqlite.sql]: No need to update schema files; migrations take care of this.

File: templates/user/settings.tmpl [hunk 21]
```diff
@@ -12,6 +12,12 @@ h3 { font-weight: normal; }
 		{{range .Flashes}}<li class="urgent">{{.}}</li>{{end}}
 	</ul>{{end}}
 
+	{{if .Suspended}}
+	<div class="alert info">
+		<p>This account is currently suspended.</p>
+		<p>Please contact the instance administrator to discuss reactivation.</p>
+	</div>
+	{{end}}
```
Comment (thebaer) [inline templates/user/settings.tmpl]: I think we should get this into a user template so we can also display it on all user backend pages (e.g. /me/c/, /me/posts/, etc.)

**Comment by thebaer** [comment]
Also resolved merge conflicts with `develop`, so be sure to pull.

File: routes.go (hunk 18 above)
Comment (robjloranger) [reply routes.go]: I went with `/admin/user/{username}/status` in aniticpation of possible future 'statuses' being added, still just toggles the status for now

**Comment by thebaer** [comment]
Made a few changes to the language, including renaming this function from **Suspending** a user to **Silencing** them. This more closely fits the function, leaves room for other kinds of moderation controls, and fits expectations admins should have with other platforms.

Since I'm going to be merging this relatively soon in order to get the release wrapped up, I'm going to note that this is experimental, and solicit bug reports and user feedback.

Some work still to do:

* Rename variables to match user-facing language
* Discuss future moderation controls

**Pull request merged**